apple/MarkdownTextInputDecoratorView.mm

```diff
@@ -51,26 +48,55 @@ - (void)didMoveToWindow {
 
 #ifdef RCT_NEW_ARCH_ENABLED
   react_native_assert([view isKindOfClass:[RCTTextInputComponentView class]] && "Previous sibling component is not an instance of RCTTextInputComponentView.");
-  _textInput = (RCTTextInputComponentView *)view;
-  _backedTextInputView = [_textInput valueForKey:@"_backedTextInputView"];
+  RCTTextInputComponentView *textInputComponentView = (RCTTextInputComponentView *)view;
+  UIView<RCTBackedTextInputViewProtocol> *backedTextInputView = [textInputComponentView valueForKey:@"_backedTextInputView"];
 #else
   react_native_assert([view isKindOfClass:[RCTBaseTextInputView class]] && "Previous sibling component is not an instance of RCTBaseTextInputView.");
-  _textInput = (RCTBaseTextInputView *)view;
-  _backedTextInputView = _textInput.backedTextInputView;
+  RCTBaseTextInputView *baseTextInputView = (RCTBaseTextInputView *)view;
+  UIView<RCTBackedTextInputViewProtocol> *backedTextInputView = baseTextInputView.backedTextInputView;
 #endif /* RCT_NEW_ARCH_ENABLED */
 
   _markdownUtils = [[RCTMarkdownUtils alloc] init];
   react_native_assert(_markdownStyle != nil);
   [_markdownUtils setMarkdownStyle:_markdownStyle];
 
-  [_textInput setMarkdownUtils:_markdownUtils];
-  if ([_backedTextInputView isKindOfClass:[RCTUITextField class]]) {
-    RCTUITextField *textField = (RCTUITextField *)_backedTextInputView;
-    _adapter = [textField valueForKey:@"textInputDelegateAdapter"];
-    [_adapter setMarkdownUtils:_markdownUtils];
-  } else if ([_backedTextInputView isKindOfClass:[RCTUITextView class]]) {
-    _textView = (RCTUITextView *)_backedTextInputView;
-    [_textView setMarkdownUtils:_markdownUtils];
+  if ([backedTextInputView isKindOfClass:[RCTUITextField class]]) {
+    _textField = (RCTUITextField *)backedTextInputView;
+    
+    // make sure `adjustsFontSizeToFitWidth` is disabled, otherwise formatting will be overwritten
+    react_native_assert(_textField.adjustsFontSizeToFitWidth == NO);
+    
+    _markdownTextFieldObserver = [[MarkdownTextFieldObserver alloc] init];
+    _markdownTextFieldObserver.markdownUtils = _markdownUtils;
+    _markdownTextFieldObserver.textField = _textField;
+    _markdownTextFieldObserver.active = YES;
+    
+    // register observers for future edits
+    [_textField addTarget:_markdownTextFieldObserver action:@selector(textFieldDidChange:) forControlEvents:UIControlEventEditingChanged];
+    [_textField addObserver:_markdownTextFieldObserver forKeyPath:@"text" options:NSKeyValueObservingOptionNew context:NULL];
+    [_textField addObserver:_markdownTextFieldObserver forKeyPath:@"attributedText" options:NSKeyValueObservingOptionNew context:NULL];
```

Comment:
Why are both (action and kv-observer) needed?

Reply:
Also, what do you think about moving registering and unregistering to the observer itself? I mean something like this:
```objc
  [_markdownTextFieldObserver observeTextField:_textField];
```

Reply:
> Why are both (action and kv-observer) needed?

Without action, the text is formatted with a small delay because of the UI-JS-UI roundtrip. The text is not formatted synchronously; it only gets updated once RN mounts the changes from JS side (because the text input in example app is controlled). The behavior is correct only when setting value during a render.

https://github.com/user-attachments/assets/3e7b2e2a-33b2-458a-82e5-037f624a7633

Without kv-observer, the opposite is happening – the formatting is visible for a short period of time before the value is updated by React Native (again, because the text input in example app is controlled).

https://github.com/user-attachments/assets/d9ba4432-b17d-4f44-8764-f46e725a45c3

With both in place, the behavior is correct:

https://github.com/user-attachments/assets/8561ecfa-2823-4095-a747-5b3f5fe40195

> Also, what do you think about moving registering and unregistering to the observer itself?

I think the current code is simpler to understand but I'll give it a try once I apply the rest of suggestions.